flink/v1.17/flink/src/main/java/org/apache/iceberg/flink/source/IcebergSource.java

```diff
@@ -453,6 +492,18 @@ public IcebergSource<T> build() {
         contextBuilder.project(FlinkSchemaUtil.convert(icebergSchema, projectedFlinkSchema));
       }
 
+      SerializableRecordEmitter<T> emitter = SerializableRecordEmitter.defaultEmitter();
+      if (watermarkColumn != null) {
```

Comment:
The focus of the feature is correct watermark generation, and we need to make sure that the watermarks are emitted in order, but this does not mean automatically that the records need to be emitted in order too. These are two different aspects of a data stream.

In case of combined splits, we do not advance the watermark, so it doesn't cause issues wrt watermark generation. The user can decide if the record out of orderness is a problem them. If they decide so, they can set the configuration, but if they have enough memory, to keep the state, they can decide that reading speed (combining files to splits) is more important than reading files in order.

Reply:
those are reasonable points. I would be interested in other people's take.

To me, it is more important to limit the out of orderliness by default. that is the whole point of watermark alignment. I would be ok to sacrifice some read throughput with smaller files. Also, with stateful applications, source read throughput is rarely the bottleneck. typical bottleneck is the stateful operator.

Reply:
At least when we use event-time alignment, the only bound we expect to be respected is the main "max out-of-orderness" constraint; the ordering of rows before the watermark is advanced isn't something we rely on. Our custom operators that use event-time won't emit results till the watermark is advanced beyond our timers anyway. So, even if the rows were received in perfect order, they'd still be buffered into state, which suggests to me there is little speed or memory benefit. Am I missing anything here?

We do use some custom process-time operators, where, in theory, less out-of-orderness would give more accurate results. But we discard the results emitted during the Iceberg backfill phase anyway, since our data are partitioned by day & the out-of-orderness allowed by ~25 hours constraint (we set it slightly above 24 hours as a precaution to avoid the aligner getting stuck w/ daily partitioned files) is too high for accurate-enough results.

I'm a bit confused on how `read.split.open-file-cost` relates to the code line this discussion is tagged at, and if maybe I'm not fully understanding.

1) Does the current code try to respect `read.split.open-file-cost` when selecting files to include in a split? IMHO, it should try to if possible, not for orderness reasons, but to avoid low throughput from potentially selecting too many small files.

2) The only other case I can think of where ordering of row reads is key is in minimizing "straggler" files that would hold up reads, i.e. if all files within a perfectly sorted daily datepartition have been read except for the one with the earliest timestamp, a max out-of-orderness of ~24 hours would mean most SplitReaders are idle, since they cannot read more than 24 hours ahead of the min timestamp. But AFAIK, the current PR's enumerator sorts the files by min timestamp & assigns them in order for this exact reason, right?

Reply:
@dchristle: The code line is mostly irrelevant for the conversation 😄

Iceberg planner collects the files which should be scanned by the execution engine, and creates `ScanTask`s from them. If a file is big, then it creates multiple splits for a given file, so multiple readers could read the given file parallel. OTOH if there are multiple small files, then it combines them to a `CombinedScanTask` which are read by a single reader - this way decreasing the number of splits, split assignments etc.

We generate a single watermark for every `ScanTask`. One `CombineScanTask` could group multiple data files, with wide range of timestamps, so generating a single watermark for it could be suboptimal. Setting `read.split.open-file-cost` could prevent the creation such `CombinedScanTask`s, and could result in better ordered input, and finer grained watermarks.

Reply:
@stevenzwu @pvary , given that the combining logic was already present before this PR and is independent, I suggest we keep the current default behaviour. (Like it's currently done in the PR)

@stevenzwu if you feel that the default combining logic should change for the Flink source, please open a separate discussion, but I personally feel that the current default is reasonable. This way Flink aligns with the default iceberg planner behaviour, but this should not be discussed further here I think.

Reply:
I am fine with tabling this as a follow-up. 

-  The default combining logic is fine for batch processing where ordering doesn't matter. for streaming read and watermark alignment, ordering is important as it affects data buffering in the flink state. That is precisely what watermark alignment intended to solve.
- Flink watermark alignment has been designed mostly with unbounded splits (like Kafka) in mind. also within an unbounded Kafka split, records are FIFO. there is a natural order within a split.

@pvary not sure every users understand the internal details. At least, we can at least document this option of disabling combining for better ordering in the doc then. Then users can make an informed choice. @dchristle can probably also chime in here.

Reply:
Watermark alignment is one of the use-cases of watermarks. Watermarks could be used for handling late data, and windowing too. So for now, added this the the javadoc of the `watermarkColumn` method:
```
     * <p>Consider setting `read.split.open-file-cost` to prevent combining small files to a single
     * split when the watermark is used for watermark alignment.
```

I hope this will help users.